cabal-dev-scripts/src/GenSPDXExc.hs

```diff
@@ -0,0 +1,126 @@
+import qualified Data.Text            as T
+import qualified Data.Text.IO         as T
+import qualified Data.Text.Lazy       as TL
+import qualified Data.Text.Lazy.IO    as TL
+import qualified Options.Applicative  as O
+import qualified Text.Microstache     as M
+
+data Opts = Opts FilePath FilePath FilePath
+
+main :: IO ()
+main = generate =<< O.execParser opts where
+    opts = O.info (O.helper <*> parser) $ mconcat
+        [ O.fullDesc
+        , O.progDesc "Generate SPDX LicenseExceptionId module"
+        ]
+
+    parser :: O.Parser Opts
+    parser = Opts <$> template <*> licenses <*> output
+
+    template = O.strArgument $ mconcat
+        [ O.metavar "SPDX.LicenseExceptionId.template.hs"
+        , O.help    "Module template file"
+        ]
+
+    licenses = O.strArgument $ mconcat
+        [ O.metavar "exceptions.json"
+        , O.help    "Exceptions JSON. https://github.com/spdx/license-list-data"
+        ]
+
+    output = O.strArgument $ mconcat
+        [ O.metavar "Output.hs"
+        , O.help    "Output file"
+        ]
+
+generate :: Opts -> IO ()
+generate (Opts tmplFile fn out) = do
+    contents <- LBS.readFile fn
+    LicenseList ls <- either fail pure $ eitherDecode contents
+    template <- M.compileMustacheFile tmplFile
+    let (ws, rendered) = generate' ls template
+    for_ ws $ putStrLn . M.displayMustacheWarning
+    TL.writeFile out (header <> "\n" <> rendered)
+    putStrLn $ "Generated file " ++ out
+
+header :: TL.Text
+header = "-- This file is generated. See Makefile's spdx rule"
+
+generate' :: [License] -> M.Template -> ([M.MustacheWarning], TL.Text)
+generate' ls template = M.renderMustacheW template $ object
+    [ "licenseIds" .= licenseIds
+    , "licenses"   .= licenseValues
+    ]
+  where
+    constructorNames :: [(Text,License)]
+    constructorNames
+        = map (\l -> (toConstructorName $ licenseId l, l))
+        $ filter (not . licenseDeprecated)
+        $ ls
+
+    licenseValues :: [Value]
+    licenseValues = flip map constructorNames $ \(c, l) -> object
+        [ "licenseCon"    .= c
+        , "licenseId"     .= textShow (licenseId l)
+        , "licenseName"   .= textShow (licenseName l)
+        ]
+
+    licenseIds :: Text
+    licenseIds = T.intercalate "\n" $ flip imap constructorNames $ \i (c, l) ->
+        let pfx = if i == 0 then "    = " else "    | "
+        in pfx <> c <> " -- ^ @" <> licenseId l <> "@, " <> licenseName l
+
+textShow :: Text -> Text
+textShow = T.pack . show
+
+toConstructorName :: Text -> Text
+toConstructorName t = t
+    & each %~ f
+    & ix 0 %~ toUpper
+    & special
+  where
+    f '.' = '_'
+    f '-' = '_'
+    f '+' = '\''
+    f c   = c
+
+    special :: Text -> Text
+    special "389_exception" = "DS389_exception"
+    special t      = t
+
+-------------------------------------------------------------------------------
+-- Licenses
+-------------------------------------------------------------------------------
+
+data License = License
```

Comment:
Shouldn't this be `LicenseException`? A bit confusing that this one and the one in `GenSPDX.hs` have the same name, but are slightly different. Maybe move these types to a common module as well?

Reply:
It should. I'll add a TODO note. I'm not keen on refactoring that too much now, as it might need bigger changes (hopefully not). It can be done after the release, as it's not part of delivery